cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1145 @@
+				f(partitionID, (*model.TableInfo)(x.target))
+			}
+		}
+		return true
+	})
+	return
+}
+
+// IterSchemas iterates all schemas in the snapshot.
+func (s *Snapshot) IterSchemas(f func(i *timodel.DBInfo)) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	s.doIterSchemas(f)
+}
+
+func (s *Snapshot) doIterSchemas(f func(i *timodel.DBInfo)) {
+	tag := negative(s.currentTs)
+	var schemaID int64 = -1
+	s.schemas.Ascend(func(i btree.Item) bool {
+		x := i.(versionedID)
+		if x.id != schemaID && x.tag >= tag {
+			schemaID = x.id
+			if x.target != nil {
+				f((*timodel.DBInfo)(x.target))
+			}
+		}
+		return true
+	})
+}
+
+// IterTableNames iterates all table names in the snapshot.
+func (s *Snapshot) IterTableNames(f func(schema int64, table string, target int64)) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	s.doIterTableNames(f)
+}
+
+func (s *Snapshot) doIterTableNames(f func(schema int64, table string, target int64)) {
+	tag := negative(s.currentTs)
+	var prefix int64 = -1
+	entity := ""
+	s.tableNameToID.Ascend(func(i btree.Item) bool {
+		x := i.(versionedEntityName)
+		if (x.prefix != prefix || x.entity != entity) && x.tag >= tag {
+			prefix = x.prefix
+			entity = x.entity
+			if x.target > 0 {
+				f(prefix, entity, x.target)
+			}
+		}
+		return true
+	})
+}
+
+// IterSchemaNames iterates all schema names in the snapshot.
+func (s *Snapshot) IterSchemaNames(f func(schema string, target int64)) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	s.doIterSchemaNames(f)
+}
+
+func (s *Snapshot) doIterSchemaNames(f func(schema string, target int64)) {
+	tag := negative(s.currentTs)
+	entity := ""
+	s.schemaNameToID.Ascend(func(i btree.Item) bool {
+		x := i.(versionedEntityName)
+		if x.entity != entity && x.tag >= tag {
+			entity = x.entity
+			if x.target > 0 {
+				f(entity, x.target)
+			}
+		}
+		return true
+	})
+}
+
+func (s *Snapshot) tablesInSchema(schema string) (tables []int64) {
+	schemaID, ok := s.doSchemaIDByName(schema)
+	if !ok {
+		return
+	}
+	start := newVersionedEntityName(schemaID, "", 0)
+	end := newVersionedEntityName(schemaID+1, "", 0)
+	tag := negative(s.currentTs)
+	currTable := ""
+	s.tableNameToID.AscendRange(start, end, func(i btree.Item) bool {
+		x := i.(versionedEntityName)
+		if x.tag >= tag && x.entity != currTable {
+			currTable = x.entity
+			if x.target > 0 {
+				tables = append(tables, x.target)
+			}
+		}
+		return true
+	})
+	return
+}
+
+// Entity(schema or table) name with finish timestamp of the associated DDL job.
+type versionedEntityName struct {
+	prefix int64 // schema ID if the entity is a table, or -1 if it's a schema.
+	entity string
+	// tag is `uint64.MAX - finish_timestamp`.
+	tag uint64
+	// the associated entity id, negative values are treated as invalid.
+	target int64
+}
+
+// ID with finish timestamp of the associated DDL job.
+type versionedID struct {
+	id  int64
+	tag uint64
+	// the associated entity pointer.
+	target unsafe.Pointer
```

Comment:
Why does it needs to be an `unsafe.Pointer`?

Reply:
Because the target can be `*model.DBInfo` or `*model.TableInfo`.

Reply:
Can it be interface{}?